Data/HashMap/Base.hs

```diff
@@ -934,6 +935,18 @@ difference a b = foldlWithKey' go empty a
                  _       -> m
 {-# INLINABLE difference #-}
 
+-- | /O(n*log m)/ Difference with a combining function. When two equal keys are
+-- encountered, the combining function is applied to the values of these keys.
+-- If it returns 'Nothing', the element is discarded (proper set difference). If
+-- it returns (@'Just' y@), the element is updated with a new value @y@.
+differenceWith :: (Eq k, Hashable k) => (v -> w -> Maybe v) -> HashMap k v -> HashMap k w -> HashMap k v
+differenceWith f a b = foldlWithKey' go empty a
+  where
+    go m k v = case lookup k b of
+                 Nothing -> insert k v m
+                 Just w  -> maybe m (\y -> insert k y m) (f v w)
```

Comment:
This bit is tricky. In the strict case you need to use `Data.HashMap.Strict.insert` to make sure the value is evaluated before being inserted in the map. You will need to duplicate this function into that module and use the right `insert`. If you look in that module you will see that happens for other functions (including `insert`) that need to force the value.